rust/signed_doc/src/lib.rs

```diff
@@ -1 +1,192 @@
 //! Catalyst documents signing crate
+#![allow(dead_code)]
+use std::{
+    convert::TryFrom,
+    fmt::{Display, Formatter},
+    sync::Arc,
+};
+
+use coset::{CborSerializable, TaggedCborSerializable};
+
+mod metadata;
+
+pub use metadata::{DocumentRef, Metadata, UuidV7};
+
+/// Catalyst Signed Document Content Encoding Key.
+const CONTENT_ENCODING_KEY: &str = "content encoding";
+/// Catalyst Signed Document Content Encoding Value.
+const CONTENT_ENCODING_VALUE: &str = "br";
+
+/// Keep all the contents private.
+/// Better even to use a structure like this.  Wrapping in an Arc means we don't have to
+/// manage the Arc anywhere else. These are likely to be large, best to have the Arc be
+/// non-optional.
+pub struct CatalystSignedDocument {
+    /// Catalyst Signed Document metadata, raw doc, with content errors.
+    inner: Arc<InnerCatalystSignedDocument>,
+}
+
+impl Display for CatalystSignedDocument {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        writeln!(f, "{}", self.inner.metadata)?;
+        writeln!(f, "JSON Payload {:#}\n", self.inner.payload)?;
+        writeln!(f, "Signatures [")?;
+        for signature in &self.inner.signatures {
+            writeln!(f, "  0x{:#}", hex::encode(signature.signature.as_slice()))?;
+        }
+        writeln!(f, "]\n")?;
+        writeln!(f, "Content Errors [")?;
+        for error in &self.inner.content_errors {
+            writeln!(f, "  {error:#}")?;
+        }
+        writeln!(f, "]")
+    }
+}
+
+#[derive(Default)]
+/// Inner type that holds the Catalyst Signed Document with parsing errors.
+struct InnerCatalystSignedDocument {
+    /// Document Metadata
+    metadata: Metadata,
+    /// JSON Payload
+    payload: serde_json::Value,
+    /// Signatures
+    signatures: Vec<coset::CoseSignature>,
+    /// Raw COSE Sign bytes
+    cose_sign: coset::CoseSign,
+    /// Content Errors found when parsing the Document
+    content_errors: Vec<String>,
+}
+
+// Do this instead of `new`  if we are converting a single parameter into a struct/type we
+// should use either `From` or `TryFrom` and reserve `new` for cases where we need
+// multiple parameters to actually create the type.  This is much more elegant to use this
+// way, in code.
+impl TryFrom<Vec<u8>> for CatalystSignedDocument {
+    type Error = anyhow::Error;
+
+    fn try_from(cose_bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        // Try reading as a tagged COSE SIGN, otherwise try reading as untagged.
+        let cose = coset::CoseSign::from_tagged_slice(&cose_bytes)
+            .or(coset::CoseSign::from_slice(&cose_bytes))
+            .map_err(|e| anyhow::anyhow!("Invalid COSE Sign document: {e}"))?;
+        let mut content_errors = Vec::new();
+        let expected_header = cose_protected_header();
+
+        if cose.protected.header.content_type != expected_header.content_type {
```

Comment:
I don't think this is flexible enough.
We should not be enforcing a particular set of header types here, we should just be requiring that `content_type` and `encoding` are both set.

We should then use these values to inform us how to handle the `payload`.